File: src/main/java/com/bcafinance/backend_saku/features/master/auditlog/service/AuditLogService.java
package com.bcafinance.backend_saku.features.master.auditlog.service;

import com.bcafinance.backend_saku.core.entity.AuditLog;
import com.bcafinance.backend_saku.core.entity.Karyawan;
import com.bcafinance.backend_saku.core.repository.AuditLogRepository;
import com.bcafinance.backend_saku.core.repository.KaryawanRepository;
import com.bcafinance.backend_saku.features.master.auditlog.dto.AuditLogItemResponse;
import com.bcafinance.backend_saku.features.master.auditlog.dto.AuditLogPageResponse;
import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Slf4j
@Service
@RequiredArgsConstructor
public class AuditLogService {

    private final AuditLogRepository auditLogRepository;
    private final KaryawanRepository karyawanRepository;

    @Transactional(readOnly = true)
    public AuditLogPageResponse getAuditLogs(
            String action,
            String entity,
            UUID karyawanId,
            String keyword,
            int page,
            int size) {

        String act = (action != null && !action.isBlank()) ? action.trim().toLowerCase() : null;
        String ent = (entity != null && !entity.isBlank()) ? entity.trim().toLowerCase() : null;
        String kw = (keyword != null && !keyword.isBlank()) ? keyword.trim().toLowerCase() : null;

        org.springframework.data.jpa.domain.Specification<AuditLog> spec = (root, query, cb) -> {
            java.util.List<jakarta.persistence.criteria.Predicate> predicates = new java.util.ArrayList<>();

            if (act != null) {
                predicates.add(cb.equal(cb.lower(root.get("action")), act));
            }
            if (ent != null) {
                predicates.add(cb.equal(cb.lower(root.get("entity")), ent));
            }
            if (karyawanId != null) {
                predicates.add(cb.equal(root.get("mstKaryawanId"), karyawanId));
            }
            if (kw != null) {
                String pattern = "%" + kw + "%";
                predicates.add(cb.or(
                        cb.like(cb.lower(root.get("description")), pattern),
                        cb.like(cb.lower(root.get("action")), pattern),
                        cb.like(cb.lower(root.get("entity")), pattern)
                ));
            }

            return cb.and(predicates.toArray(new jakarta.persistence.criteria.Predicate[0]));
        };

        Pageable pageable = PageRequest.of(Math.max(0, page), Math.max(1, size), Sort.by(Sort.Direction.DESC, "createdDate"));
        Page<AuditLog> logPage = auditLogRepository.findAll(spec, pageable);

        List<AuditLogItemResponse> items = logPage.getContent().stream()
                .map(this::mapToItemResponse)
                .toList();

        return AuditLogPageResponse.builder()
                .totalItems(logPage.getTotalElements())
                .totalPages(logPage.getTotalPages())
                .currentPage(logPage.getNumber())
                .pageSize(logPage.getSize())
                .logs(items)
                .build();
    }


    @Transactional
    public void recordLog(UUID karyawanId, String action, String entity, Integer entityId, String description) {
        if (karyawanId == null) {
            return;
        }
        try {
            AuditLog logEntry = new AuditLog();
            logEntry.setId(UUID.randomUUID());
            logEntry.setMstKaryawanId(karyawanId);
            logEntry.setAction(action != null ? truncate(action.toUpperCase(), 20) : "ACTIVITY");
            logEntry.setEntity(entity != null ? truncate(entity.toUpperCase(), 20) : "GENERAL");
            logEntry.setEntityId(entityId != null ? entityId : 0);
            logEntry.setDescription(description != null ? truncate(description, 225) : "");
            logEntry.setCreatedDate(LocalDateTime.now());

            auditLogRepository.save(logEntry);
            log.info("Audit log recorded: [{}] {} by karyawan {}", action, description, karyawanId);
        } catch (Exception e) {
            log.error("Failed to record audit log: {}", e.getMessage(), e);
        }
    }

    @Transactional
    public void recordLog(UUID karyawanId, String action, String entity, String description) {
        recordLog(karyawanId, action, entity, 0, description);
    }

    private String truncate(String val, int maxLen) {
        if (val == null) return "";
        return val.length() > maxLen ? val.substring(0, maxLen) : val;
    }


    private AuditLogItemResponse mapToItemResponse(AuditLog a) {
        AuditLogItemResponse.KaryawanAuditInfo karyawanInfo = null;

        if (a.getMstKaryawanId() != null) {
            Optional<Karyawan> karyawanOpt = karyawanRepository.findById(a.getMstKaryawanId());
            if (karyawanOpt.isPresent()) {
                Karyawan k = karyawanOpt.get();
                String roleName = k.getRole() != null ? k.getRole().getNama() : "-";
                String cabangName = k.getCabang() != null ? k.getCabang().getNama() : "-";

                karyawanInfo = AuditLogItemResponse.KaryawanAuditInfo.builder()
                        .id(k.getId())
                        .nama(k.getNama())
                        .email(k.getEmail())
                        .username(k.getUsername())
                        .role(roleName)
                        .cabang(cabangName)
                        .build();
            }
        }

        return AuditLogItemResponse.builder()
                .id(a.getId())
                .action(a.getAction())
                .entity(a.getEntity())
                .entityId(a.getEntityId())
                .description(a.getDescription())
                .createdDate(a.getCreatedDate())
                .karyawan(karyawanInfo)
                .build();
    }
}

File: src/app/core/services/audit-log.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, timeout, map, catchError, of } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse } from '../models/auth.models';
import { AuditLog, AuditLogPageResponse } from '../models/audit-log.model';

@Injectable({
  providedIn: 'root',
})
export class AuditLogService {
  private http = inject(HttpClient);
  private readonly baseUrl = `${environment.apiUrl}/master/audit-log`;
  private readonly fallbackUrl = `${environment.apiUrl}/audit-log`;

  getAuditLogs(params?: {
    action?: string;
    entity?: string;
    karyawanId?: string;
    keyword?: string;
    page?: number;
    size?: number;
  }): Observable<AuditLogPageResponse> {
    let httpParams = new HttpParams();
    if (params) {
      if (params.page !== undefined) httpParams = httpParams.set('page', String(params.page));
      if (params.size !== undefined) httpParams = httpParams.set('size', String(params.size));
      if (params.action) httpParams = httpParams.set('action', params.action);
      if (params.entity) httpParams = httpParams.set('entity', params.entity);
      if (params.karyawanId) httpParams = httpParams.set('karyawanId', params.karyawanId);
      if (params.keyword) httpParams = httpParams.set('keyword', params.keyword);
    }

    return this.http
      .get<ApiResponse<AuditLogPageResponse> | any>(this.baseUrl, { params: httpParams })
      .pipe(
        timeout(6000),
        map((res: any) => this.parsePageResponse(res, params?.size || 10)),
        catchError(() => {
          return this.http
            .get<any>(this.fallbackUrl, { params: httpParams })
            .pipe(
              timeout(6000),
              map((res: any) => this.parsePageResponse(res, params?.size || 10)),
              catchError(() =>
                of({
                  logs: [],
                  content: [],
                  totalElements: 0,
                  totalItems: 0,
                  totalPages: 0,
                  currentPage: 0,
                  pageSize: params?.size || 10,
                })
              )
            );
        })
      );
  }

  private parsePageResponse(res: any, defaultSize: number): AuditLogPageResponse {
    let items: AuditLog[] = [];
    let totalElements = 0;
    let totalPages = 1;
    let currentPage = 0;
    let pageSize = defaultSize;

    if (res && res.data && typeof res.data === 'object') {
      if (Array.isArray(res.data.logs)) {
        items = res.data.logs;
        totalElements = res.data.totalItems ?? res.data.totalElements ?? items.length;
        totalPages = res.data.totalPages ?? Math.ceil(totalElements / defaultSize);
        currentPage = res.data.currentPage ?? res.data.page ?? 0;
        pageSize = res.data.pageSize ?? res.data.size ?? defaultSize;
      } else if (Array.isArray(res.data.content)) {
        items = res.data.content;
        totalElements = res.data.totalElements ?? res.data.totalItems ?? items.length;
        totalPages = res.data.totalPages ?? Math.ceil(totalElements / defaultSize);
        currentPage = res.data.currentPage ?? res.data.page ?? 0;
        pageSize = res.data.pageSize ?? res.data.size ?? defaultSize;
      } else if (Array.isArray(res.data)) {
        items = res.data;
        totalElements = items.length;
        totalPages = Math.ceil(items.length / defaultSize) || 1;
      }
    } else if (res && Array.isArray(res.logs)) {
      items = res.logs;
      totalElements = res.totalItems ?? res.totalElements ?? items.length;
      totalPages = res.totalPages ?? Math.ceil(totalElements / defaultSize);
      currentPage = res.currentPage ?? res.page ?? 0;
      pageSize = res.pageSize ?? res.size ?? defaultSize;
    } else if (res && Array.isArray(res.content)) {
      items = res.content;
      totalElements = res.totalElements ?? items.length;
      totalPages = res.totalPages ?? Math.ceil(totalElements / defaultSize);
      currentPage = res.currentPage ?? res.page ?? 0;
      pageSize = res.pageSize ?? res.size ?? defaultSize;
    } else if (Array.isArray(res)) {
      items = res;
      totalElements = items.length;
      totalPages = Math.ceil(items.length / defaultSize) || 1;
    }

    const normalizedItems = items.map((item) => ({
      ...item,
      karyawanNama: item.karyawan?.nama || item.karyawanNama || item.username || '-',
      username: item.karyawan?.username || item.username || '-',
      role: item.karyawan?.role || item.role || '-',
      timestamp: item.createdDate || item.timestamp,
    }));

    return {
      logs: normalizedItems,
      content: normalizedItems,
      totalItems: totalElements || normalizedItems.length,
      totalElements: totalElements || normalizedItems.length,
      totalPages: totalPages || 1,
      currentPage,
      pageSize,
    };
  }
}
